### ISSUE TITLE
species level classfication

### ISSUE THREAD

**Issue by yhuang15** [benjjneb/dada2#542]
Hi,  I wonder if it would be ok to re-format those files for species assignment (e.g. silva_species_assignment_v132.fa.gz) into ones with 7 taxonomy level and simply use them for assignTaxonomy() to get species assignment using the RDP classifier algorithm?   What could be potential problems?

Thanks!
Yanmei

**Comment by benjjneb** [comment]
The fuzzy matching + bootstrap approach used by the RDP classifier isn't appropriate for the species level for 16S gene fragments. It's too permissive and will give lots of false positives. It's effective at higher taxonomic levels, but just wasn't designed for assigning to the species level from partial 16S.

`assignSpecies` uses unambiguous exact matching, which recent publications have indicated is the only appropriate way to assign species IDs to partial 16S (at least when using general purpose databases).

**Comment by yhuang15** [comment]
On Wed, Aug 15, 2018 at 9:10 AM, Benjamin Callahan <notifications@github.com
> wrote:

> The fuzzy matching + bootstrap approach used by the RDP classifier isn't
> appropriate for the species level for 16S gene fragments. It's too
> permissive and will give lots of false positives. It's effective at higher
> taxonomic levels, but just wasn't designed for assigning to the species
> level from partial 16S.
>
>
Am I on the right track to think that the reason that RDP classifier works
well for higher taxonomic level is because at higher levels there are lots
of sequences for the classifier to "learn" from, whereas at species level
there are too few sequences in most cases?  Thus, if I have a specialized
database that has lots of different sequences for each individual species,
then I could potential do species level classification using the
classifier, correct?

**Comment by benjjneb** [comment]
> if I have a specialized database that has lots of different sequences for each individual species, then I could potential do species level classification using the classifier, correct?

Maybe. That would be better at least. The other issue is that fuzzy matching just usually isn't appropriate for species assignment. See for example this recent paper: https://www.ncbi.nlm.nih.gov/pubmed/29506021

**Comment by yhuang15** [comment]
Thanks for the article. It was a good read. I totally agree that 97%
identity is way too low for species assignment, to a point of silliness,
and that optimal thresholds are data-dependent.  However, I still think
100% match is too stringent and in my experiences caused a lot of reads
being unassigned due to the exact match not existing in database, residual
PCR errors not being corrected by DADA2 (those that occurred early in PCR
cycles), etc.  Also, it is a pity not to take advantage of the ability of
the RDP classifier to "learn" from the training dataset and assign more
weights to more informative Kmers.  For example, A and B are both 99.5%
identical to C, but the mismatches  between A and C are in position that
are critical for defining species C, whereas mismatches between B and C are
in some unimportant places, then only B, but not A should be classified as
species C, despite both having 99.5% identity to C.  With sufficient amount
of training data, RDP classifier might be able to resolve that.

**Comment by benjjneb** [comment]
I agree with what you are saying, and I will assign 16S fragments (at least tentatively) to the species level via a slightly fuzzy approach of BLASTing sequences against nt in environments where I also have some domain knowledge about the species expected there. I don't trust anything other than exact matching as an automated method right now though, for partial 16S. But if you put aside my opinions, you can use `assignTaxonomy` in the way you suggested in the first post (i.e. reformatting to include species level).

You may also be interested in the recently introduced IDTaxa classifier. [The paper just came out](https://microbiomejournal.biomedcentral.com/articles/10.1186/s40168-018-0521-5) and reports a real advance over the longtime standard set by the naive Bayesian classifier. 

And it's available through the R/Bioconductor DECIPHER package! https://bioconductor.org/packages/release/bioc/html/DECIPHER.html

My impression is that @digitalwright (the developer) is very skeptical of sequence-level assignment in general 16S data though, so I think species-level assignment was only evaluated on ITS data.

**Comment by yhuang15** [comment]
This is exactly what I hoped. I am working with a group of microbiologists
who are experts in oral and nasal bacteria and I felt I could use their
domain knowledge to help construct a training dataset specific for bacteria
of the oral and nasal sites and use the RDP classifier down to species
level.  Will do a lot of testings to see if that really is possible but it
helps to know your opinion first so that I know what to expect.  Thanks!
Also thanks for the info about IDTaxa.  Will try that out.

On Wed, Aug 15, 2018 at 4:55 PM, Benjamin Callahan <notifications@github.com

**Comment by TseWue** [comment]
Dear @yhuang15 and Benjamin,

I just read this thread and I may have something to add to the discussion (a bit late I know). I am currently working on nasal samples in which we are focusing on the genus of Staphylococci. For this, we amplified a specific generegion which is said to be promising for exact species detection within the Staphylococcal family. I have put together a taxonomy and species assignment database of over 9000 sequences (for inclusion and exclusion of Staphylococcus spp.) and yet, I am not able to assign all the SVs, even though when blasting I get a clear result, sometimes even down to strain level. However, since the addSpecies() function only allows 100% matches that sort of requires my database to include every SNP possible at each position in the gene. This is impossible. Especially, since most of the SNPs have never been described. Therefore I would like to ask, if you know a workaround to allow, let's say for 1 or 2 mismatches? Or would you recommend to use another program to do so?

Thanks for your help.
Tse

**Comment by benjjneb** [comment]
You could try to use the `assignTaxonomy` method, with your reference database that has data assigned down to the species level. You'll need to reformat the id lines though, into the Rank1;Rank2;Rank3... format expected by `assignTaxonomy`.

`assignTaxonomy` doesn't require exact matching, and if your locus is more variable than the species level (i.e. each species has multiple if not many different sequence variants associated with it) then it might work well for your purposes.

I'd recommend adding a few outgroup sequences as well into the training fasta though, to keep the algorithm honest.

**Comment by TseWue** [comment]
Dear Benjamin

Thanks for the tip! I reformatted the db and tried it but unfortunately, it was not possible for the gene I am using. Everything classified as S. aureus, since aureus sequences were the most abundant in the db. Anyhow, I'll keep this tweak in mind for other projects! :-) Right now I try to work around this issue by using pyhlogenetic clustering, i.e. making a simple tree and seeing where the SVs classify within the rest of the db sequences. 

Thanks for your superfast reply, appreciate it!
Best wishes,
Tse

**Comment by yhuang15** [comment]
Dear Tse,

I was working on developing a method for species/subgenus level
classification using the RDP classifier with enhanced training datasets.
Our training set includes skin and nasal microbiome, thus is expected to
include Staphylococci.  The manuscript is currently under review at
Microbiome.  However, I have moved to a new position a year ago and no
longer working on microbiome.  You could contact Dr. Katherine Lemon of The
Forsyth Institute to obtain the training set and the improved method if
interested.

Best,
Yanmei
